{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "19684ed4-fa3e-480d-a2b5-1796c079753b",
   "metadata": {},
   "source": [
    "# Drug-Seq 数据分析"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac93e694-9584-40c7-b243-d06ba16ee551",
   "metadata": {},
   "source": [
    "## rawdata 预处理\n",
    "\n",
    "使用 `drug_seq_v2.py` 脚本，对 fastq 文件进行预处理"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1bc3f6f-0a7d-455c-8a99-45088f9e7650",
   "metadata": {},
   "outputs": [],
   "source": [
    "#drug seq data preprocess\n",
    "import os\n",
    "import sys\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import json\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import gzip\n",
    "import pysam\n",
    "\n",
    "\n",
    "######################################################\n",
    "##1. 读入fastq1，获取barcode、UMI\n",
    "##2. 对数据进行质控与可视化\n",
    "######################################################\n",
    "def get_fastq1(barcode,fastq1):\n",
    "    R1_reads={}\n",
    "    with gzip.open(fastq1, 'rt') as f:\n",
    "        for line in f:\n",
    "            sequence_identifier = line.strip()  # 序列标识符，通常是 '@' 开头的行\n",
    "            sequence_ID=re.sub(\" [0-9]*\\\\/[0-9]*\",\"\",sequence_identifier)\n",
    "            if sequence_identifier.startswith('@'):\n",
    "                sequence = next(f).strip()  # 序列行\n",
    "                barcode_id=sequence[:10]\n",
    "                UMI=sequence[-10:]\n",
    "                _ = next(f).strip()\n",
    "                quality_scores = next(f).strip()  # 质量分数行\n",
    "                R1_reads[sequence_ID]=barcode_id+\"_\"+UMI\n",
    "\n",
    "    barcodes=get_barcode_library(barcode)\n",
    "    rs={} #用于存储解析后的barcode_UMI，仅限有效的barcodes\n",
    "    for k,v in R1_reads.items():\n",
    "        r1=re.sub(\"\\\\_.*\",\"\",v)\n",
    "        if r1 in barcodes:\n",
    "            r2=re.sub(\".*\\\\_(.*)\",\"\\\\1\",v)\n",
    "            if r1 in rs:\n",
    "                if r2 in rs[r1]:\n",
    "                    rs[r1][r2] +=1\n",
    "                else:\n",
    "                    rs[r1][r2]=1\n",
    "            else:\n",
    "                rs[r1]={r2:1}\n",
    "\n",
    "    #每个barcode下的UMI，转录本数目\n",
    "    barcode_umi_library_size={}\n",
    "    for k,v in rs.items():\n",
    "        barcode_umi_library_size[k]=len(v)        \n",
    "    df=pd.DataFrame.from_dict(barcode_umi_library_size, orient='index')\n",
    "    df.columns=['N']\n",
    "    df.to_csv(\"UMI_counts_in_Barcodes.csv\")\n",
    "\n",
    "    plt.hist(df[['N']], bins=30, edgecolor='black')  # bins参数控制直方图的条形数，edgecolor设置条形边缘颜色\n",
    "    plt.title('N of transcripts in 360 wells')# 添加标题和标签\n",
    "    plt.xlabel('Value')\n",
    "    plt.ylabel('Frequency')\n",
    "    plt.savefig(\"UMI_counts_distribution.png\")# 显示直方图\n",
    "    plt.close()\n",
    "\n",
    "\n",
    "    #每个wells中，UMI>1的转录本个数\n",
    "    barcode_umi_dupliates={}\n",
    "    for k,v in rs.items():\n",
    "        barcode_umi_dupliates[k]=sum(1 for value in v.values() if value > 1)/len(v)\n",
    "    df=pd.DataFrame.from_dict(barcode_umi_dupliates, orient='index')\n",
    "    df.columns=['Rate']\n",
    "    df.to_csv(\"UMI_duplicates.csv\")\n",
    "    \n",
    "    plt.hist(df[['Rate']], bins=30, edgecolor='black')  # bins参数控制直方图的条形数，edgecolor设置条形边缘颜色\n",
    "    plt.title('UMI duplicate rates')# 添加标题和标签\n",
    "    plt.xlabel('Value')\n",
    "    plt.ylabel('Frequency')\n",
    "    plt.savefig(\"UMI_duplicates.png\")# 显示直方图\n",
    "    plt.close()\n",
    "    \n",
    "    return(R1_reads)\n",
    "\n",
    "\n",
    "######################################################\n",
    "##获取有效barcode，减少不必要的检索\n",
    "##\n",
    "######################################################\n",
    "def get_barcode_library(barcode):\n",
    "    barcode_id=pd.read_csv(barcode)\n",
    "    barcodes=list(barcode_id['barcode_ID'])\n",
    "    return barcodes\n",
    "\n",
    "######################################################\n",
    "##将fastq2比对到参考基因组\n",
    "##\n",
    "######################################################\n",
    "def align_to_reference(sample,fastq2):\n",
    "    _cmd={}\n",
    "    _cmd['STAR']= '''/OIU/sge_software/STAR/STAR-2.4.2a/bin/Linux_x86_64/STAR \\\n",
    "        --genomeDir /OIU/database_NGS/hg38/reference/STAR_242a_index \\\n",
    "        --sjdbGTFfile /OIU/database_NGS/hg38/GTF/gencode.v45.primary_assembly.annotation.gtf \\\n",
    "        --limitBAMsortRAM 40000000000 \\\n",
    "        --runThreadN 4 \\\n",
    "        --limitIObufferSize 500000000 \\\n",
    "        --outFilterType BySJout \\\n",
    "        --outFilterMismatchNmax 999 \\\n",
    "        --outFilterMismatchNoverLmax 0.04 \\\n",
    "        --outFilterMultimapNmax 20 \\\n",
    "        --outFilterMatchNminOverLread 0.66 \\\n",
    "        --outFilterIntronMotifs None \\\n",
    "        --outSJfilterReads All \\\n",
    "        --outSAMtype BAM SortedByCoordinate \\\n",
    "        --outSAMunmapped Within \\\n",
    "        --outSAMstrandField intronMotif \\\n",
    "        --outSAMattrRGline ID:{prefix} SM:{prefix} PL:ILLUMINA \\\n",
    "        --alignSJoverhangMin 8 \\\n",
    "        --alignSJDBoverhangMin 1 \\\n",
    "        --alignIntronMin 20 \\\n",
    "        --alignIntronMax 1000000 \\\n",
    "        --alignMatesGapMax 1000000 \\\n",
    "        --chimSegmentMin 15 \\\n",
    "        --chimJunctionOverhangMin 15 \\\n",
    "        --chimScoreMin 0 \\\n",
    "        --chimScoreDropMax 20 \\\n",
    "        --chimScoreSeparation 10 \\\n",
    "        --chimScoreJunctionNonGTAG -1 \\\n",
    "        --quantMode TranscriptomeSAM \\\n",
    "        --quantTranscriptomeBan IndelSoftclipSingleend \\\n",
    "        --outReadsUnmapped Fastx \\\n",
    "        --readFilesIn  {R2} \\\n",
    "        --readFilesCommand zcat \\\n",
    "        --outFileNamePrefix  {prefix}.'''.format(prefix=sample,R2=fastq2)\n",
    "        \n",
    "    _cmd['sort']='''/OIU/sge_software/samtools/currentVersion/samtools sort {prefix}.Aligned.toTranscriptome.out.bam > {prefix}.Aligned.toTranscriptome.out_sort.bam'''.format(prefix=sample)\n",
    "    _cmd['index']='''/OIU/sge_software/samtools/currentVersion/samtools index {prefix}.Aligned.toTranscriptome.out_sort.bam'''.format(prefix=sample)\n",
    "    os.system(_cmd['STAR'])\n",
    "    os.system(_cmd['sort'])\n",
    "    os.system(_cmd['index'])\n",
    "\n",
    "\n",
    "######################################################\n",
    "##解析bam文件，获取reads和genes\n",
    "##\n",
    "######################################################\n",
    "def get_reads_genes_from_BAM(sample):\n",
    "    reads_genes={}#包括： reads-> genes\n",
    "    with pysam.AlignmentFile(\"{}.Aligned.toTranscriptome.out_sort.bam\".format(sample), \"r\") as bamfile:\n",
    "        for read in bamfile.fetch():\n",
    "            if not read.is_unmapped and read.query_name.startswith('SRR'):\n",
    "                read_records=read.to_dict()\n",
    "                if read_records['name'] in reads_genes:\n",
    "                    reads_genes[read_records['name']].append(read_records['ref_name'])\n",
    "                else:\n",
    "                    reads_genes[read_records['name']]=[read_records['ref_name']]   \n",
    "\n",
    "    return reads_genes\n",
    "\n",
    "######################################################\n",
    "##联合R1,R2分析结果，将readsID 置换为 barcode_ID 和 UMI\n",
    "##\n",
    "######################################################\n",
    "def exchange_readsID_with_barcodeUMI(barcode,R1_reads,reads_genes):\n",
    "    tmp={}\n",
    "    barcodes=get_barcode_library(barcode)\n",
    "    for k,v in reads_genes.items():\n",
    "        r1=re.sub(\"\\\\_.*\",\"\",R1_reads[\"@\"+k])\n",
    "        if r1 in barcodes:#即为有效barcode\n",
    "            if R1_reads[\"@\"+k] in tmp:\n",
    "                #print(tmp[R1_reads[\"@\"+k]])\n",
    "                tmp[R1_reads[\"@\"+k]].update({g:np.round(1/len(v),2) for g in v})\n",
    "                #print(tmp[R1_reads[\"@\"+k]])\n",
    "                #i +=1\n",
    "                #print(i)#难道只是PCR重复带来的UMI重复？\n",
    "                #print(\"--------------------------\")\n",
    "            else:\n",
    "                tmp[R1_reads[\"@\"+k]]={g:np.round(1/len(v),2) for g in v}\n",
    "    return(tmp)\n",
    "\n",
    "######################################################\n",
    "##定量，在不同barcode中，定量转录本的表达\n",
    "##\n",
    "######################################################\n",
    "def counts_UMI_for_transcripts_in_each_barcodes(tmp):\n",
    "    barcode_expression={}\n",
    "    for k,v in tmp.items():\n",
    "        r1=re.sub(\"\\\\_.*\",\"\",k)#获取barcode 序列\n",
    "        for g,c in v.items():\n",
    "            if r1 in barcode_expression:\n",
    "                if g in barcode_expression[r1]:\n",
    "                    barcode_expression[r1][g] += c\n",
    "                else:\n",
    "                    barcode_expression[r1][g]=c\n",
    "            else:\n",
    "                barcode_expression[r1] = {g:c}   \n",
    "    return barcode_expression\n",
    "\n",
    "\n",
    "######################################################\n",
    "##定量，在不同barcode中，定量基因水平的表达\n",
    "##\n",
    "######################################################\n",
    "def counts_UMI_in_gene_level(transcripts_counts):\n",
    "    barcode_genes={}\n",
    "    gene_transcript=get_transciprts_genes_from_GTF()\n",
    "    for barcode,transripts in transcripts_counts.items():\n",
    "        barcode_genes[barcode]={}\n",
    "        for transriptsID,counts in transripts.items():\n",
    "            if gene_transcript[transriptsID] in barcode_genes[barcode]:\n",
    "                barcode_genes[barcode][gene_transcript[transriptsID]] += np.round(counts,0) #不同基因转录本数据相加\n",
    "            else:\n",
    "                barcode_genes[barcode][gene_transcript[transriptsID]] = np.round(counts,0)  \n",
    "    barcode_expression_df=pd.DataFrame.from_dict(barcode_genes)\n",
    "    return barcode_expression_df\n",
    "\n",
    "def get_transciprts_genes_from_GTF():\n",
    "    GTF=\"/OIU/database_NGS/hg38/GTF/gencode.v45.primary_assembly.annotation.gtf\"\n",
    "    gene_transcript={}\n",
    "    with open(GTF,\"r\") as f:\n",
    "        fs=f.readlines()\n",
    "        for rs in fs:\n",
    "            if re.search(\"^chr\",rs):\n",
    "                records=re.split(\"\\t\",rs)\n",
    "                if records[2]==\"transcript\":\n",
    "                    gs=re.split(\";\",records[8])\n",
    "                    gs_transcript_id=re.split(\"\\s\",gs[1])[2].strip('\\\\\"\"')\n",
    "                    gs_genesymbol=re.split(\"\\s\",gs[3])[2].strip('\\\\\"\"')\n",
    "                    gene_transcript[gs_transcript_id]=gs_genesymbol\n",
    "    return gene_transcript\n",
    "\n",
    "#计算信息\n",
    "if __name__ == '__main__':\n",
    "    if len(sys.argv) != 5:\n",
    "        print(\"Usage: python drugseq.py barcode fastq1 fastq2 sample\")\n",
    "    else:\n",
    "        barcode=sys.argv[1]\n",
    "        fq1=sys.argv[2]\n",
    "        fq2=sys.argv[3]\n",
    "        sample=sys.argv[4]\n",
    "        #step1\n",
    "        R1_reads=get_fastq1(barcode, fq1) #barcode,fastq1\n",
    "        #step2\n",
    "        align_to_reference(sample, fq2) ##barcode,fastq2, 比对\n",
    "        #step3\n",
    "        reads_genes=get_reads_genes_from_BAM(sample)\n",
    "        #step4\n",
    "        tmp=exchange_readsID_with_barcodeUMI(barcode,R1_reads,reads_genes)\n",
    "        transcripts_counts=counts_UMI_for_transcripts_in_each_barcodes(tmp)#转录本水平获取表达数据\n",
    "        count_matrix=counts_UMI_in_gene_level(transcripts_counts)#基因水平获取表达数据\n",
    "        count_matrix.to_csv(sample+\"_count_matrix_gene_level.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d2b96d1-8328-4e6b-b2ab-e52dbb396083",
   "metadata": {},
   "source": [
    "## matrix分析\n",
    "\n",
    "1. 文章中matrix的分析\n",
    "\n",
    "2. 复现文章中的上游数据处理"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b69e77bf-c499-468a-be1f-43942697a53b",
   "metadata": {},
   "source": [
    "### 对一个样本进行drug-seq分析\n",
    "\n",
    "主要有：\n",
    "\n",
    "1. 聚类\n",
    "2. 差异分析\n",
    "3. 结果可视化等"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "eb77765f-8d03-4e9d-a512-7386774acd0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(\"D:\\\\Cell_panel_screening\\\\DRUG-Seq\\\\GSM5357044\")\n",
    "#360 wells, 5w+ genes\n",
    "m=load(\"GSM5357044_flowcell_4000_UMI_decode_VH02001612.RData\")\n",
    "\n",
    "library(data.table)\n",
    "library(dplyr)\n",
    "library(ggplot2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6de2ef35-a074-46a8-8c6d-fd0dd30422cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.\n",
      "\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAIAAAByhViMAAAABmJLR0QA/wD/AP+gvaeTAAAg\nAElEQVR4nO3dT2zbd/348Y/tj/8kaTe2cSBBWzh0I5qQQATt0F6RrA4lQpVQERLabTnkqzIO\n7QFNVJM4lcNYpQy1Oe0YhKYphk2B46DaoREHQEzrJNRJxCcQabpSmjj+Hj6/+ZdvW7tpavvj\nvPZ4HCb748x55523nWc/H9ufQrvdTgAAOPyKeQ8AAID+EHYAAEEIOwCAIIQdAEAQwg4AIAhh\nBwAQhLADAAhC2AEABJHmPYB92dra2tnZyXsU/VepVKrV6tbWVt4DGVG1Wq1Wq+3u7t64cSPv\nsYyoarVaLpdv3ryZ90BG1NjYWLVabbVaHmXd1Gq1Uqn06aef5j2QETU+Pl6pVHZ2djzKuhkb\nGysUCrdu3cp7ICMqW0Lb29v9fZQ98cQT3W46HGG3u7vbarXyHkX/Zaf9CPmj9UW73S4UCokp\n6s4S6s0SeiBLqDdL6IGyKTI/PQx5CTkUCwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQd\nAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCKGRngB5xOzs7pVIp71EMRKFwOH4FecnOnWyK\nerCEesjWT2IJ9WQJ9WAJPZBn6d4GsYRarVaapt1u7XrDSLl9+3ar1cp7FP1XqVQqlcrNmzfz\nHsiIqlartVptd3d3a2sr77GMqEqlUi6XP/3007wHMqJqtVq1Wm21Wh5l3VSr1TRNLaFuxsbG\nKpXKzs6OKeqmVqsVi8Vbt27lPZARNT4+Xi6X+76EHn/88W43HY6wa7Va29vbeY+i/0qlUrvd\nDvmj9UW5XE6SxBT1kKap+emhUqkkllBP5XJ5d3fX/HRTrVYTS6in7FFmfrrZ3d3N/ju0KfIa\nOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhh\nBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQad4D4PPlzJkzfbmfixcv9uV+ACAS\ne+wAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7AIAg\nhB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCE\nsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQ\ndgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDC\nDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELY\nAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7\nAIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgiHQI\n32NpaWltbS1JkpmZmQsXLjxwOwAABzDwPXZLS0vPPPPM6urq6urq9PT0uXPnsu2NRuP69evZ\n9uzLBj0SAIDYBht2zWZzbW3tW9/6Vnb11KlTH374YbPZTJJkeXn59OnT2fbTp09nu+4AADiw\nwYbdxsbGzMzM5ORkdnVycnJ1dXVycjJru6mpqWz77OxskiTr6+sDHQwAQGyDfY3dxsbG9PT0\n+vr6a6+9lm3JDrxubGwkSdIJvnt99NFHv/71rztXv/e97z399NMDHWouSqVSsVg8cuRI3gM5\nfExaJk1TS6iHNE2TJCmVSqaoG0uot3K5nFhCPaVpWigUzE832bNQmqZ9nKLd3d1e37Ff3+a+\nPvnkk+wYa9Zz6+vr586d6/Y+iY2NjWzXXZIk//jHP95+++3OTd/+9refffbZgQ41R7VaLe8h\nPNgPfvCDvIfwfxyKSRsas9FboVAwRb2Zn96KxaIp6q1UKuU9hJHW3yXUarV63Drwd8XOzMws\nLi5ml2dnZ1977bVuh1w7R2aTJHnyySdfeOGFztXx8fHt7e2BjjMXxWKxWCzu7OzkPZDDJ+R6\nOABLqLdsp3i73TZF3RSLxUKh0PvvxOeZJfRAWdJZQt0MYgnt7u72KOnBht0zzzxz3+1ZwzWb\nzW5HY7/+9a+/+eabnaubm5ubm5uDGGG+arVarVYL+aMNmknLjI2NVSoVs9HNxMTE2NhYq9Uy\nRd2Mj4+naXrjxo28BzKijhw5UqvVdnZ2LKFuJiYmisXi1tZW3gMZUUePHq1Wq9vb2/19lFWr\n1W43DfbNE1NTU/e+3XVqairrueyVdslnb5voHIcFAOAABht2s7OzMzMzjUYju9poNDpvkq3X\n6ysrK9n2lZWVer0+0JEAAIQ38A8ovnDhwvvvvz8/Pz8/P//+++933jmxuLg4PT2dbZ+enu68\nDg8AgIMZxinFur0NdnFxUc8BAPTLwPfYAQAwHMIOACAIYQcAEISwAwAIQtgBAAQh7AAAghB2\nAABBCDsAgCCEHQBAEMIOACAIYQcAEISwAwAIQtgBAAQh7AAAghB2AABBCDsAgCCEHQBAEMIO\nACAIYQcAEISwAwAIQtgBAAQh7AAAghB2AABBCDsAgCCEHQBAEMIOACAIYQcAEISwAwAIQtgB\nAAQh7AAAghB2AABBCDsAgCCEHQBAEMIOACAIYQcAEISwAwAIQtgBAAQh7AAAghB2AABBCDsA\ngCCEHQBAEMIOACAIYQcAEISwAwAIQtgBAAQh7AAAghB2AABBCDsAgCCEHQBAEMIOACAIYQcA\nEISwAwAIQtgBAAQh7AAAghB2AABBCDsAgCCEHQBAEMIOACAIYQcAEISwAwAIQtgBAAQh7AAA\nghB2AABBCDsAgCCEHQBAEMIOACAIYQcAEISwAwAIQtgBAAQh7AAAghB2AABBCDsAgCCEHQBA\nEMIOACAIYQcAEISwAwAIQtgBAAQh7AAAghB2AABBCDsAgCCEHQBAEMIOACAIYQcAEISwAwAI\nQtgBAAQh7AAAghB2AABBCDsAgCCEHQBAEMIOACAIYQcAEISwAwAIQtgBAAQh7AAAghB2AABB\nCDsAgCDSvAewL8VisVQq5T2K/isUCoVCIeSPNmgmLWMJ9VYoFLL/mqJuLKHeLKEHsoR6G8QS\narfbPW49HGE3NjaWpodjqAfwxBNP5D2Ew8ek7WU2eiuVSqaoN/PTW5qmpqi3SqWS9xBGWrlc\n7uMSarVaPW49HLV069at7e3tvEfRf9VqtVarbW5u5j2Qw+ef//xn3kMYCbVarVKp3LhxI++B\njKjx8fGxsbFWq/Xvf/8777GMqOyfzVtbW3kPZERNTEzUarWdnR1P1N2Mj48Xi8WbN2/mPZAR\ndeTIkWq1eufOnf4+yp566qluNx2OsGu32713PB5qgX+0wTFpe5mN3mI/gfSF+enNEnog8/NA\nQ5sib54AAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHs\nAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQd\nAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLAD\nAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYA\nAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4A\nIAhhBwAQhLADAAhC2AEABCHsAACCSPMeAAN35syZvIcAAAyDPXYAAEEIOwCAIIQdAEAQwg4A\nIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEA\nBCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCA\nIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIIjhhV2z\n2Zyfn282m50tS0tL8/Pz8/Pz586dG9owAACiGl7Yvf7663uvNhqN69evr66urq6uJkmytLQ0\ntJEAAIQ0pLBrNBp3bVleXj59+nR2+fTp02tra8MZCQBAVMMIu2azuby8/OMf/3jvliRJpqam\nsquzs7NJkqyvrw9hMAAAUaVD+B6vv/76+fPn927Z2NhIkmRycrLb/3L16tU33nijc/Xs2bPP\nP//84EaYl2KxWCgUvvCFL+Q9kMPHpGUsod6KxWKSJKVSyRR1Ywn1li2hNE1NUTeWUG+lUilJ\nknK53Mcp2t3d7XHrwMOu0WhMT0/Pzs7ufdvEfW1sbGS77pIk2dra+tvf/ta56fbt22k6jAbN\nReAfbXBM2l5mo7dCoWCKejM/vVlCD2R+euvvEmq1Wj1uHexvIjsIm7094oE6R2aTJPnyl798\n6tSpztUnnnji9u3b/R9f3kqlUqlUunPnTt4DOXxCrocDSNO0WCxaQt2kaZqmabvd/u9//5v3\nWEaUJdRbuVwulUq7u7umqJs0TQuFwvb2dt4DGVGDWEK7u7vj4+Pdbh1s2F29ejVJkvn5+c6W\nhYWF8+fPZw3XbDa7HY197rnnfvKTn3Subm5u3rx5c6BDzUWtVqvVaiF/tEEzaZmxsbFKpWI2\nupmYmEjTtNVqmaJuxsfH0zQ1P90cOXKkVCpZQj1MTEwUi0Xz083Ro0dLpdLOzk5/pyi3sJub\nm5ubm8suN5vNhYWFS5cudWJuY2Mju5y9baJzHBYAgAPI7cwT9Xp9ZWUlu7yyslKv1/MaCQBA\nDLmF3eLi4vT0dHbmienp6cXFxbxGAgAQw/DexjI5OXnXuygWFxf1HABAv+S2xw4AgP4SdgAA\nQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAg\nCGEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAE\nkeY9ADiIM2fO9OV+Ll682Jf7AYBRYI8dAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACC2G/Y\nnThx4sSJE/fdXigU+jokAAAO4gGfY/fxxx9nF65cubL3aubatWvZdgAAcveAsHv22Wd7XM0c\nP368nyMCAOBAHhB2165du3btWpIkL774YpIk77777r1fc/LkyUGMDACAh/KAsDt27NixY8eS\nJHn55ZcTDQcAMML2e67YS5cuDXQcAAA8Ih93AgAQxEOEXfbJJvc1uPEBALBP+z0Uu7CwkH2y\nyS9+8YvnnntukEMCAOAg9ht2ly9fTpLk2rVr2XspAAAYNQ/3GjtVBwAwsvYbdtmnEN915gkA\nAEbHfsPurbfeSpLkpZde0nYAAKNpv6+xy04mduXKlfueVazdbvdzUAAAPLz9hl125gkAAEaW\nM08AAAThzBMAAEEIOwCAIPZ7KLb3ecO8eQIAIHf7Dbt33333ri3vvPPO5cuXnWEMAGBE7Dfs\nTp48ee+W559//pVXXrm3+QAAGL5Heo3dj370oyRJfvazn/VpMAAAHFwf3jxx5cqVR78TAAAe\n0SOF3RtvvJF8dhpZAADy1Yd3xb766qt9GgwAAAf3qKcUO3v27LFjx/o3HgAADsgpxQAAgnDm\nCQCAIPa7xy7z3nvvvfPOO3/5y1+SJPna17723e9+997PtwMAIBcPEXYnTpzY+8kmV65cuXz5\n8vHjx//4xz8OYGAAADyc/R6KfeONN65cuXL8+PFr16612+12u33t2rXjx49fuXIl+9ATAADy\ntd+w+9WvfpUkyVtvvdV5D+yxY8feeuutzk0AAORrv2GXHYS965NNsqvOPAEAMAoe7l2xH3/8\ncY+rAADkaL9hl31A8W9/+9u9G7Or3T67GACAYdpv2J09ezZJkldeeeXEiRPvvffee++9d+LE\niVdeeaVzEwAA+drvx50cO3bs2rVrL7300pUrV1588cVs4/Hjx/e+nQIAgBw9xOfYHTt2zEfW\nAQCMLKcUAwAI4pHCzrtiAQBGx0OE3cLCQqFQ2Lvl2WefPXHiRL+HBADAQTzEKcWyM8Pu3fjy\nyy9fuXJlYWFhAAMDAODh7Dfssk82yc4h1nHp0qUkSS5fvtz3YQEA8LAe7jV2PtkEAGBk7Tfs\nsoOw77333t6N2dW7js8CAJCL/Ybdq6++miTJiy++uLCwkJ15YmFhIfuk4uwmAADytd8PKD55\n8uS77777s5/97PLly50X1R0/fvzVV189efLkwIYHAMB+PcSZJ06ePKnhAABGljNPAAAEIewA\nAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABDEQ3yOHUN25syZvIcAABwm9tgBAAQh7AAAghB2\nAABBCDsAgCCEHQBAEMIOACAIYQcAEISwAwAIQtgBAARxOM48USqV8h7CQJRKpUKhUC6X8x7I\n59dhn/xisWgJ9VAsFpMkMUU9FIvFYrFofrqxhB7IEuotW0LDnKLDEXa1Wi1q2xUKhcceeyzv\nUXx+BZh8S6iHQqGQJEmpVDJFPVhCPWRLKE1TU9RNNkXmp5tBLKFWq9Xj1sMRdp9++un29nbe\no+i/Wq1Wq9X+/e9/5z2Qz69//vOfeQ/hkYyNjVUqlc3NzbwHMqImJibGxsZ2dnY8yroZHx9P\n0/TGjRt5D2REHTlypFarbW9ve5R1MzExUSwWt7a28h7IiDp69Gi1Wr1z505/H2Vf/OIXu93k\nNXYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQ\nwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC\n2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEI\nOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABJHmPQDI05kzZx79Ti5evPjodwIAj84eOwCA\nIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQ\nhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACC\nEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQ\nwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC\n2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIIg07wHAoXfmzJm+3M/Fixf7cj8AfG7ZYwcA\nEISwAwAIQtgBAAQh7AAAghB2AABBDONdsfPz853Lq6urnctLS0tra2tJkszMzFy4cGEIIwEA\nCGzge+zm5+fr9frq6urq6mq9Xj937ly2vdFoXL9+PdueJMnS0tKgRwIAENtgw259fT1JklOn\nTmVXT5069eGHHzabzSRJlpeXT58+nW0/ffp0tusOAIADG2zYzc7Orq6uTk5O3rU9a7upqanO\nlyWfVSAAAAcz1DNPvP322zMzM5OTk1nD3Rt8HX/4wx9++tOfdq7+/Oc//+Y3vzmMIQ5doVB4\n6qmn8h4FI+FgK8ES6qFQKCRJkqapKerBEuohW0LlctkU9VAoFCqVSt6jGFHZEqpUKn1cQq1W\nq8etwwu7RqOxtrZ26dKlbl+wsbGR7bpLkmR7e/vGjRudm1qtVjY1IQX+0XgoB14JltADmaLe\nzM8DmaLezM8D9XGKet/VkMKu0WgsLy+fP3++x166zpHZJEm++tWv/uQnP+lc/dKXvnTz5s3B\nDjEP5XK5XC7funUr74EwEg6wyMvlcpqm//nPfwYxngCq1Wq5XN7d3fUo66ZSqRSLxdu3b+c9\nkBGVLaFWq+VR1k21Wi0UCpZQN7VaLU3T/i6hdrt99OjRbrcOI+yyjzW5dOlSp+qyhms2m906\nb2pqqvOWiyRJNjc3oy6aUqkU9UfjYR1gJRQKBX+VeyiVSlnYmaJuisVimqbmp5s0TS2h3kql\nkmehHrJ/frdarf5OUY+wG/jHnWRHYO96C0V2eWNjI7uaveSucxwWAIADGGzYNZvN5eXl+76u\nrl6vr6ysZJdXVlbq9fpARwIAEN5gD8VevXo1SZKFhYW9G8+fPz87O7u4uLi0tJSdlKJery8u\nLg50JAAA4Q027Obm5ubm5rrduri4qOcAAPpl4K+xAwBgOIQdAEAQwg4AIAhhBwAQhLADAAhC\n2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEI\nOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhh\nBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHs\nAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQd\nAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLAD\nAAhC2AEABCHsAACCEHYAAEEIOwCAINK8BwD8P2fOnHn0O7l48eKj3wkAh5Q9dgAAQQg7AIAg\nhB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCE\nsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQ\ndgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACCLNewD7Ui6X0/RwDPWhpGlaLBbHxsbyHghx\nWE57Zc8bHmU9pGlaKpXMTzeW0AOlaVooFMxPN6VSKftvH6eo3W73uPVw1FKpVCoWA+5cLBaL\nhUKhUqnkPRDisJz2yp5SPcp6KBaLxWLR/HST/ekxRT1kjzLz080gltDu7m6PWw9H2N2+fXt7\nezvvUfRfrVar1Wqbm5t5D4Q4LKe9JiYmxsbGWq2WaelmfHw8TdMbN27kPZARdeTIkVqttrOz\nYwl1MzExUSwWt7a28h7IiDp69Gi1Wt3e3u7vo6xarXa7KeBuMACAzydhBwAQhLADAAhC2AEA\nBCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCA\nIIQdAEAQwg4AIAhhBwAQRJr3AIB+OnPmTF/u5+LFi325HwCGyR47AIAghB0AQBDCDgAgCGEH\nABCEsAMACELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBBOKTYQ/TqtEwDA/tljBwAQhLADAAhC\n2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhDNPAAPRx/OvXLx4sV93xedBv9ae\nhcdhZI8dAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAin\nFANgJPTxNHTwuWWPHQBAEMIOACAIYQcAEISwAwAIQtgBAAQh7AAAghB2AABBCDsAgCCEHQBA\nEM48AcTXx1MaXLx4sV93BdB39tgBAAQh7AAAghB2AABBCDsAgCCEHQBAEMIOACAIYQcAEISw\nAwAIQtgBAAThzBMAPKo+ntsjmFGbGadOCc8eOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEA\nBCHsAACCEHYAAEEIOwCAIIQdAEAQTikG3MeonQdpdIzUzDg91ECN1O8a9skeOwCAIIQdAEAQ\nwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIJx54v/wOePAIeIpi4fVlzXTl1Oe\nhFy9o3AyGHvsAACCEHYAAEEIOwCAIIQdAEAQwg4AIIg83xW7tLS0traWJMnMzMyFCxdyHAkA\nQAC57bFrNBrXr19fXV1dXV1NkmRpaSmvkQAAxJBb2C0vL58+fTq7fPr06WzXHQAAB5bPodhm\ns5kkydTUVHZ1dnY2SZL19fXsQpIk//nPf/71r391vr5arZZKpaEPExgJB374FwqFzn+Bftn7\nkCwUCoVCwd/ozL3z0HkW6uMUtdvtHrfmE3YbGxtJkkxOTnb7gg8++ODs2bOdq2+++eYLL7ww\njJEBo+eJJ554lP/dnxzor3sfkpVKJZeRjJpuT1blcvkRn8f2arVaPW4doVOKbWxsdPbYAXT8\n4Ac/yHsIwP/nITnKRijsOkdmkyT5+te//uabb3auPv3005ubm0MYwy9/+cshfJeOSqVSqVRu\n3rw5zG96iFSr1Vqttru7u7W1lfdYRlS1Wk3T9NNPP817ICOqVqtVq9VWq+VR1k2tViuVSpZQ\nN2NjY5VKZWdnxxR1U6vVisXirVu38h7ISLi3VcbHx8vlct+X0OOPP97tpnzCLmu4ZrPZ7Wjs\nk08+uffY6+bm5vb29pAGN0SlUqndbof80fqiXC4nSWKKekjTtFQqmZ9ussNDllAP5XK5UCiY\nn26q1WpiCfWUPcrMTze7u7vZf4c2Rfm8KzbrueyVdkmSrK+vJ5+9hQIAgIPJ7eNO6vX6yspK\ndnllZaVer+c1EgCAGHILu8XFxenp6fn5+fn5+enp6cXFxbxGAgAQQ55vnlhcXNRzAAD9ktse\nOwAA+kvYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACFG9/woAAAVgSURB\nVELYAQAEIewAAIIQdgAAQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAA\nQQg7AIAghB0AQBDCDgAgCGEHABCEsAMACELYAQAEIewAAIIQdgAAQaR5D+BzbXt7u91u5z2K\n0fX+++9/8MEHjz322A9/+MO8xzKi7ty502q18h7F6Pr973//pz/96amnnvr+97+f91hG1J07\nd3Z2dvIexej6zW9+89e//nVycvLUqVN5j2VE/fe//y0UCnmPYnS9/fbbH3300Ve+8pXvfOc7\nw/mOhyPsHn/88byHMEBHjx7Newgj6pNPPmk0GlNTUz/+8Y/zHguH0t///vdGo/Hss8/+z//8\nT95j4VD66KOPGo3GN77xjZdffjnvsXAo/fnPf/7d73534sSJl156aTjf0aFYAIAghB0AQBDC\njtFVrVYfe+yxI0eO5D0QDqtsCU1MTOQ9EA6rWq322GOPjY+P5z0QDquxsbEhL6GCF+8DAMRg\njx0AQBDCDgAgCGEHABDE4fgcOw67ZrO5sLBw6dKlycnJbMvS0tLa2lqSJDMzMxcuXOh85aC3\nc+icO3fuww8/TIa+VCyhAObn5/c+7WRbOpdXV1c7ly0n7uveJZTXUtn/ErLHjmF4/fXX915t\nNBrXr19fXV3NnliXlpaGs51D59y5c0mSZL/K6enp7GpiCbEP9/7W5ufn6/V69mut1+uWE73d\n+5taWlp65plnRv0ZqQ0Dtrq6evbs2bm5uY2NjWzL3Nzc1atXs8tXr16dm5sbznYOl42Njb3L\nZu9VS4je5j7TWT/Z79FyYp/uXUKH5RnJHjsGq9lsLi8v7z0nWLPZTJJkamoquzo7O5skyfr6\n+qC3D/THZJgsIXpbX1+v1+uXLl3au3F2dnZ1dXXvMbWM5cS97ruENjY2ZmZmOktocnIyW1Gj\ntoSEHYP1+uuvnz9/fu+WjY2NJEnufXod9HYOncnJyZmZmc5x/Lfffjt7VrWE6G12dnZxcbH3\n11hO9HDfJbSxsTE9Pb2+vj7/mc72ZJSWkLBjgBqNxvT0dPbPi96yhTv87Yy4CxcuTE9PZ8+h\n169f7/GSYUuI/Ws0Gmtra3uPJNzFcuJen3zyydra2gcffJC91u38+fOd19jdK8clJOwYlOwg\n7AP/3Zzp7GQe8nZGXPZv4s5Lhve+pfEulhD71Gg0lpeXz58/32MXiOXEfc3MzHT+qM3Ozn74\n4YfdDonmuISEHYNy9erVJEmyfS0LCwtJkiwsLKyvr2fLMXvRwF6D3s6hkz1jdp5Gs/0rzWbT\nEuLAlpaWlpeXL1261DmSYDmxT88888z09PS920dtCQk7BmVubm71M9lLULMn0+xfyZ3dyNkf\n7yFsH8KPzHBYQhxMdgT2rrdQWE7s09TUVPZJcndtHLUlJOzIQb1eX1lZyS6vrKzU6/XhbOdw\nyZ65Op/Y1Hm1e2IJ8fCyF4fc9T7HjOXEfszOzs7MzDQajexqo9EYzWekQrvdftSfFR6kx5kn\n6vX63tfhDXo7h07ndXXdPofdEuK+7nrayV5ad9fXnD9/vvPvB8uJu9z7lyvZx7lwcl9Cwg4A\nIAiHYgEAghB2AABBCDsAgCCEHQBAEMIOACAIYQcAEISwAwAIQtgBAAQh7AAAghB2AABBCDsA\ngCCEHQBAEMIOACAIYQcAEISwAwAIQtgBAAQh7AAAghB2AABBCDsAgCCEHcC+vPHGG4VC4eOP\nP857IABdCTsAgCAK7XY77zEAANAH9tgBAAQh7AAAghB2AABBCDsAgCCEHQBAEMIOACAIYQcA\nEISwAwAIQtgBAAQh7AAAghB2AABB/C/zjGoAY5KGYgAAAABJRU5ErkJggg==",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#各样本文库大小\n",
    "colSums(UMI_decode) %>% \n",
    "  as.data.frame() %>% \n",
    "  ggplot(aes(x=.))+\n",
    "  geom_histogram()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "866e147f-05c6-4ede-828f-13da29936e64",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 15 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>compound</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>(S)-Crizotinib   </td><td>24</td></tr>\n",
       "\t<tr><td>AZD8055          </td><td>24</td></tr>\n",
       "\t<tr><td>BLU9931          </td><td>24</td></tr>\n",
       "\t<tr><td>Brusatol         </td><td>24</td></tr>\n",
       "\t<tr><td>BTdCPU           </td><td>24</td></tr>\n",
       "\t<tr><td>CHIR118637       </td><td>24</td></tr>\n",
       "\t<tr><td>Cmp_334          </td><td>24</td></tr>\n",
       "\t<tr><td>Cmp_341          </td><td>24</td></tr>\n",
       "\t<tr><td>CPI-203          </td><td>24</td></tr>\n",
       "\t<tr><td>Dilazep          </td><td>24</td></tr>\n",
       "\t<tr><td>DMSO             </td><td>24</td></tr>\n",
       "\t<tr><td>Fedratinib       </td><td>24</td></tr>\n",
       "\t<tr><td>Homoharringtonine</td><td>24</td></tr>\n",
       "\t<tr><td>NVS-SM2          </td><td>24</td></tr>\n",
       "\t<tr><td>Triptolide       </td><td>24</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 15 × 2\n",
       "\\begin{tabular}{ll}\n",
       " compound & n\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t (S)-Crizotinib    & 24\\\\\n",
       "\t AZD8055           & 24\\\\\n",
       "\t BLU9931           & 24\\\\\n",
       "\t Brusatol          & 24\\\\\n",
       "\t BTdCPU            & 24\\\\\n",
       "\t CHIR118637        & 24\\\\\n",
       "\t Cmp\\_334           & 24\\\\\n",
       "\t Cmp\\_341           & 24\\\\\n",
       "\t CPI-203           & 24\\\\\n",
       "\t Dilazep           & 24\\\\\n",
       "\t DMSO              & 24\\\\\n",
       "\t Fedratinib        & 24\\\\\n",
       "\t Homoharringtonine & 24\\\\\n",
       "\t NVS-SM2           & 24\\\\\n",
       "\t Triptolide        & 24\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 15 × 2\n",
       "\n",
       "| compound &lt;chr&gt; | n &lt;int&gt; |\n",
       "|---|---|\n",
       "| (S)-Crizotinib    | 24 |\n",
       "| AZD8055           | 24 |\n",
       "| BLU9931           | 24 |\n",
       "| Brusatol          | 24 |\n",
       "| BTdCPU            | 24 |\n",
       "| CHIR118637        | 24 |\n",
       "| Cmp_334           | 24 |\n",
       "| Cmp_341           | 24 |\n",
       "| CPI-203           | 24 |\n",
       "| Dilazep           | 24 |\n",
       "| DMSO              | 24 |\n",
       "| Fedratinib        | 24 |\n",
       "| Homoharringtonine | 24 |\n",
       "| NVS-SM2           | 24 |\n",
       "| Triptolide        | 24 |\n",
       "\n"
      ],
      "text/plain": [
       "   compound          n \n",
       "1  (S)-Crizotinib    24\n",
       "2  AZD8055           24\n",
       "3  BLU9931           24\n",
       "4  Brusatol          24\n",
       "5  BTdCPU            24\n",
       "6  CHIR118637        24\n",
       "7  Cmp_334           24\n",
       "8  Cmp_341           24\n",
       "9  CPI-203           24\n",
       "10 Dilazep           24\n",
       "11 DMSO              24\n",
       "12 Fedratinib        24\n",
       "13 Homoharringtonine 24\n",
       "14 NVS-SM2           24\n",
       "15 Triptolide        24"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "meta=fread(\"cb1c00920_si_001/File_S1_metadata.csv\") %>% \n",
    "  filter(plate_barcode==\"VH02001612\")\n",
    "#3个sectors，8个 dose，所以每个药物共有24例数据\n",
    "meta %>% count(compound)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb89094e-5f56-4985-a824-56724c3647ad",
   "metadata": {},
   "source": [
    "#### UMAP聚类"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19b06a0f-8dac-43a7-b196-492f94c19e15",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(limma)\n",
    "library(edgeR)\n",
    "library(umap)\n",
    "library(plotly)\n",
    "\n",
    "tmp=UMI_decode[rowSums(UMI_decode)>0,]\n",
    "d<-DGEList(counts=tmp)\n",
    "d <- calcNormFactors(d, method=\"TMM\")\n",
    "mat_ump_wbatchE<-log2(edgeR::cpm(d,  normalized.lib.sizes=T,log=F)+1)#scale到一个相对可比的范围    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e295ed06-0783-4add-a460-b93d787168c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts=mat_ump_wbatchE\n",
    "pca<-prcomp(t(counts), scale = T, center = T)\n",
    "rownames(pca$x)  <-  colnames(counts)\n",
    "umap  <- umap(pca$x[,1:20],method = \"naive\",n_neighbors=15)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f5a3712-8ec7-4c95-88de-2a9f6deaeb8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#umap结果可视化\n",
    "colnames(umap$layout)<-c('umap1','umap2')\n",
    "color=meta %>% select(Sample,plate_barcode,treatment,batch,treatment_dose,compound) %>% as.data.frame()\n",
    "rownames(color)=meta %>% mutate(ID=paste0(plate_barcode,\"_\",well_index)) %>% pull(ID)\n",
    "\n",
    "tmp=cbind(\n",
    "  umap$layout,\n",
    "  color[rownames(umap$layout),]\n",
    ")\n",
    "\n",
    "p=tmp %>% plot_ly(x = ~umap1, y = ~umap2,\n",
    "                color = ~compound,type = \"scatter\",mode = 'markers', \n",
    "                size = ~treatment_dose,fill = ~'',\n",
    "                text = ~compound)\n",
    "htmlwidgets::saveWidget(as_widget(p), \"tmp_500plotly.html\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "358c337a-7677-4e23-8f18-ea59c04a2d07",
   "metadata": {},
   "source": [
    "+ UMAP分组可视化"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15127840-278e-4ae5-9698-7359255e1301",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp_labels=tmp %>%  distinct(.,Sample,.keep_all = T)\n",
    "library(RColorBrewer)\n",
    "compounds=unique(tmp$compound)\n",
    "compounds_colors=colorRampPalette(brewer.pal(8, \"Set2\"))(length(compounds))\n",
    "names(compounds_colors)=compounds\n",
    "\n",
    "p=tmp %>% \n",
    "  ggplot(aes(x=umap1,y=umap2))+\n",
    "  geom_point(\n",
    "    aes(colour=compound,size=treatment_dose)\n",
    "  )+\n",
    "  scale_colour_manual(\n",
    "    values = c(compounds_colors)\n",
    "  )+\n",
    "  ggrepel::geom_text_repel(\n",
    "    aes(x=umap1,y=umap2,label=compound),\n",
    "    max.overlaps = 100,\n",
    "    data=tmp_labels\n",
    "  )+\n",
    "  facet_wrap(~treatment_dose)+\n",
    "  theme_bw(base_size = 20)\n",
    "\n",
    "ggsave('umap.png',p,width = 15,height = 15)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19cb6850-103f-4743-b28a-9bed9a796533",
   "metadata": {},
   "source": [
    "+ 基本可以复现文章中的分析结果"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "780c9601-c17b-4ac8-9429-7ff80750a92a",
   "metadata": {},
   "source": [
    "#### Heatmap可视化"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e77cea2-1e57-43d2-9936-3abd75373bc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#heatmap非监督聚类，看一看不同药物之间的clusters\n",
    "summary(\n",
    "  apply(mat_ump_wbatchE,\n",
    "        1,\n",
    "        sd)\n",
    ")\n",
    "x1=mat_ump_wbatchE[ apply(mat_ump_wbatchE,\n",
    "                          1,\n",
    "                          sd) >0.9,]\n",
    "\n",
    "library(pheatmap)\n",
    "#选择样本进行可视化\n",
    "select_samples=meta %>% filter(treatment_dose>3) %>% mutate(ID=paste0(plate_barcode,\"_\",well_index)) %>% \n",
    "  select(c(ID,compound,treatment_dose))\n",
    "\n",
    "annotation_col=data.frame(\n",
    "  Compounds=select_samples$compound,\n",
    "  Dose=select_samples$treatment_dose\n",
    ")\n",
    "rownames(annotation_col)=select_samples$ID\n",
    "\n",
    "pheatmap(\n",
    "  x1[,select_samples$ID],\n",
    "  cluster_cols = T,\n",
    "  cluster_rows = F,\n",
    "  annotation_col = annotation_col,\n",
    "  annotation_colors = list(\n",
    "    Compounds=compounds_colors\n",
    "  ),\n",
    "  show_rownames = F,\n",
    "  show_colnames = F,\n",
    "  scale = 'none',\n",
    "  filename = 'heatmap.png'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abdc1fc4-2ad0-4b73-bae3-c63965962ebc",
   "metadata": {},
   "source": [
    "+ 有待进一步优化"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76f88c0b-79d9-4d3d-b1fd-1c07fec80551",
   "metadata": {},
   "source": [
    "### 类单细胞分析\n",
    "\n",
    "比较有趣的是，可以进行类单细胞分析，结合matrix以及meta文件\n",
    "\n",
    "1. 选择dose为10的样本进行类单细胞分析\n",
    "2. 选择dose在3以上的样本进行类单细胞分析"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d22b2b3-f619-43c9-ad61-0e53d30d5ce6",
   "metadata": {},
   "source": [
    "#### 差异分析\n",
    "\n",
    "each compounds vs 所有的DMSO,所以一共有14组compounds分析结果"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2516389-89b0-4e17-9a65-39cd0d75a3f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(Seurat)\n",
    "select_samples=meta %>% filter(treatment_dose==10 | compound=='DMSO') %>% mutate(ID=paste0(plate_barcode,\"_\",well_index)) %>% \n",
    "  select(c(ID,compound,treatment_dose))\n",
    "select_counts=UMI_decode[,select_samples$ID]\n",
    "\n",
    "obj=CreateSeuratObject(\n",
    "  counts = select_counts,\n",
    "  project= \"Drug_seq\"\n",
    ")\n",
    "obj@meta.data=cbind(obj@meta.data,select_samples)\n",
    "\n",
    "obj$Group=as.factor(obj$compound)\n",
    "Idents(obj)='compound'\n",
    "de.res <- list()\n",
    "for (i in levels(obj$Group)) {\n",
    "  print(i)\n",
    "  if(i!=\"DMSO\"){\n",
    "    DEG <- FindMarkers(obj, \n",
    "                       #第一组哪些样本\n",
    "                       ident.1 = i, \n",
    "                       #第2组哪些样本\n",
    "                       ident.2 = 'DMSO',\n",
    "                       test.use = \"DESeq2\",\n",
    "                       verbose = FALSE)\n",
    "    DEG$compound <- i \n",
    "    DEG$Gene <- rownames(DEG)\n",
    "    rownames(DEG) <- NULL\n",
    "    DEG <- DEG[-1,]\n",
    "    de.res[[i]] <- DEG\n",
    "  }\n",
    "}\n",
    "out <- do.call(rbind, lapply(de.res, data.frame))\n",
    "#out$Compare <- paste(levels(combined$Group)[1], \"_vs_\", levels(combined$Group)[2], sep = \"\")\n",
    "#out <- out[,c(\"CellType\", \"Gene\", \"avg_log2FC\", \"pct.1\", \"pct.2\", \"p_val\", \"p_val_adj\", \"Compare\")]\n",
    "openxlsx::write.xlsx(out, \"compoundsType_degs_deseq2.xlsx\", rowNames = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3c48491-9afb-4f1b-9c2d-e8f44cb3a4af",
   "metadata": {},
   "source": [
    "#### 差异分析结果可视化"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7817483-4302-4542-8f6f-48e5e2089ed2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#部分结果可视化\n",
    "heatmap_markers=out %>%  filter(p_val_adj<0.05 & abs(avg_logFC)>2) %>% \n",
    "  group_by(compound) %>% \n",
    "  slice_max(.,avg_logFC,n=10) %>% \n",
    "  pull(Gene)\n",
    "\n",
    "\n",
    "DoHeatmap(object = obj1, features = heatmap_markers,slot='counts')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbe2f862-1ab2-46a5-9105-eca7419edbea",
   "metadata": {},
   "source": [
    "#### TSNE/UMAP 聚类"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b459d5ee-0baf-4e93-a99a-a570a5fcb34f",
   "metadata": {},
   "outputs": [],
   "source": [
    "select_samples=meta %>% mutate(ID=paste0(plate_barcode,\"_\",well_index)) %>% \n",
    "  select(c(ID,compound,treatment_dose)) #选择所有样本进行分析\n",
    "select_counts=UMI_decode[,select_samples$ID]\n",
    "obj=CreateSeuratObject(\n",
    "  counts = select_counts,\n",
    "  project= \"Drug_seq\"\n",
    ")\n",
    "obj@meta.data=cbind(obj@meta.data,select_samples) #更新meta文件"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08a0f9cb-1ecf-4cba-973c-ab3fd918045e",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj1 <- NormalizeData(object = obj)\n",
    "obj1 <- FindVariableFeatures(object = obj1)\n",
    "obj1 <- ScaleData(object = obj1)\n",
    "obj1 <- RunPCA(object = obj1)\n",
    "obj1 <- FindNeighbors(object = obj1, dims = 1:30)\n",
    "obj1 <- FindClusters(object = obj1)\n",
    "obj1 <- RunUMAP(object = obj1, dims = 1:30)\n",
    "obj1 <- RunTSNE(object = obj1, dims = 1:30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9775ec96-24b2-4e6b-b70d-6f82cbfdcead",
   "metadata": {},
   "outputs": [],
   "source": [
    "Idents(obj1)=\"compound\"\n",
    "p1=DimPlot(obj1, reduction = \"umap\", label = T,label.size = 5,\n",
    "          repel = T) +NoLegend()\n",
    "p2=DimPlot(obj1, reduction = \"tsne\", label = T,label.size = 5,\n",
    "           repel = T) +NoLegend()\n",
    "ps=ggpubr::ggarrange(p1,p2,nrow = 1)\n",
    "ggsave(\"umap_tsne_seurat.png\",ps, width =12, height = 6)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f11a91f-66f8-4911-a7a7-f6ab8033ea3c",
   "metadata": {},
   "source": [
    "### 差异基因的功能分析\n",
    "\n",
    "主要是 `GSEA` 分析"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54a0902f-51bb-47fb-b037-89ca4355e0d2",
   "metadata": {},
   "source": [
    "#### GO分析 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09b9650c-1273-43e3-835a-254fec5a0ef9",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(clusterProfiler)\n",
    "library(org.Hs.eg.db)\n",
    "library(ggplot2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61952544-1b13-4590-8780-b81db4a19a18",
   "metadata": {},
   "outputs": [],
   "source": [
    "#对不同药物作用后差异基因进行功能分析\n",
    "out=readxl::read_xlsx(\"compoundsType_degs_deseq2.xlsx\") %>% \n",
    "  mutate(GeneSymbol=gsub('\\\\,grch.*','',Gene)) %>% select(-c(Gene))\n",
    "#对不同compounds差异基因进行功能分析\n",
    "getEntrezID=function(geneSymbol){\n",
    "  library(easyConvert)\n",
    "  tmp=easyConvert(species = \"HUMAN\",queryList = geneSymbol,queryType = \"SYMBOL\")\n",
    "  return(as.character(tmp$ENTREZID))\n",
    "}\n",
    "compounds=unique(out$compound)\n",
    "ups=lapply(\n",
    "  compounds,\n",
    "  FUN=function(x){\n",
    "    gs=out %>% filter(compound==x) %>% filter(p_val_adj<0.05 & abs(avg_logFC) >0.5) %>% pull(GeneSymbol)\n",
    "    gs_list=getEntrezID(gs)\n",
    "    return(gs_list)\n",
    "  }\n",
    ")\n",
    "names(ups)=compounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58895580-a65c-41e1-8ab9-280ea18855bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "ck_up <- compareCluster(geneCluster = ups, \n",
    "                        fun = \"enrichGO\",\n",
    "                        OrgDb=\"org.Hs.eg.db\",pvalueCutoff=1,qvalueCutoff  = 1)\n",
    "#将EntrezID，转换为Gene sybmol\n",
    "ck_up_geneSymbol <- setReadable(ck_up, 'org.Hs.eg.db', 'ENTREZID')\n",
    "dotplot(ck_up_geneSymbol)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29429642-da1e-4efb-815c-0b86f4665435",
   "metadata": {},
   "source": [
    "+ 结论：\n",
    "\n",
    "这种比较分析， 没有给出更多的信息，尝试进行 GSEA分析吧"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f958300a-483f-498c-9f97-58bd4e133040",
   "metadata": {},
   "source": [
    "#### C2\n",
    "\n",
    "包括：\n",
    "\n",
    "+ CGP, Chemical and genetic perturbations\n",
    "+ CP, Canonical pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41fda56e-d48b-427c-b7f7-46bfe672e2e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#进行尝试进行GSEA分析吧\n",
    "library(msigdbr)\n",
    "library(GSEABase)\n",
    "library(easyConvert)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc88cdb0-e8f9-4ddc-933e-6bcf3553ab60",
   "metadata": {},
   "outputs": [],
   "source": [
    "b = msigdbr(species = \"Homo sapiens\", category = \"C2\") %>%\n",
    "  dplyr::select(gs_name, entrez_gene)\n",
    "saveRDS(b,'gsea_c2.rds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4aaf5246-0360-4458-a479-13bff42f632b",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_results=lapply(\n",
    "  compounds,\n",
    "  FUN=function(x){\n",
    "    print(x)\n",
    "    gs=out %>% filter(compound==x) %>% filter(p_val_adj<0.05) %>% dplyr::select(c('GeneSymbol','avg_logFC'))\n",
    "    \n",
    "    if(nrow(gs)>1){ #有些组可能相比DMSO，没有差异基因\n",
    "      a1 = gs$avg_logFC\n",
    "      a2 = easyConvert(\n",
    "        species = \"HUMAN\",\n",
    "        queryList = gs$GeneSymbol,\n",
    "        queryType = \"SYMBOL\"\n",
    "      )\n",
    "      names(a1) = a2$ENTREZID\n",
    "      b1 = a1[order(a1, decreasing = T)]\n",
    "      \n",
    "      gsearesults = GSEA(b1, TERM2GENE = b,pvalueCutoff=1)\n",
    "      return(gsearesults)\n",
    "    }else{\n",
    "      gsearesults=list()\n",
    "      return(gsearesults)\n",
    "    }\n",
    "  }\n",
    ")\n",
    "names(gsea_results)=compounds\n",
    "#saveRDS(gsea_results,'gsea_results.rds')\n",
    "saveRDS(gsea_results,'gsea_c2_results.rds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81d00e82-be14-4004-862c-a90b9493a67e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#不同化合物gsea分析结果以list形式存储\n",
    "gsea_results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "deb74678-c086-46c9-83e6-1bc452ad3798",
   "metadata": {},
   "source": [
    "+ 结果可视化"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1366900-1a01-41d8-8ed8-950e52f766d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#GSEA结果可视化\n",
    "ps_gseas=list()\n",
    "ps_gseas=lapply(\n",
    "  compounds,\n",
    "  FUN = function(x){\n",
    "    print(x)\n",
    "    tmp=gsea_results[[x]]\n",
    "    if(length(tmp)){\n",
    "      p=tmp@result %>% mutate(Description = stringr::str_to_title(gsub(\"_\", \" \", Description))) %>%\n",
    "        mutate(Description = reorder(Description, setSize)) %>% head(10) %>%\n",
    "        ggplot(aes(\n",
    "          x = enrichmentScore,\n",
    "          y = Description,\n",
    "          size = setSize,\n",
    "          color = pvalue\n",
    "        )) +\n",
    "        geom_point() +\n",
    "        scale_size(range = c(1, 5), name = \"Size\") +\n",
    "        scale_color_gradient(low = \"blue\", high = \"red\") +\n",
    "        xlab(\"EScore\") + ylab(\"\") + labs(title=x)+\n",
    "        theme(axis.text.y = element_text(size = 10, color = \"black\"))+\n",
    "        scale_y_discrete(labels = function(x) stringr::str_wrap(x, width = 25))+\n",
    "        theme_bw(base_size = 15)\n",
    "      \n",
    "    }else{\n",
    "      p=ggplot()\n",
    "    }\n",
    "    return(p)\n",
    "  }\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a76ab47-f19b-4574-a825-bbebeb639d62",
   "metadata": {},
   "outputs": [],
   "source": [
    "#保存\n",
    "ggsave(\n",
    "  filename = \"GSEA_c2_plots.pdf\", \n",
    "  plot = gridExtra::marrangeGrob(ps_gseas, nrow=1, ncol=2), \n",
    "  width = 12, height = 6\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e46d258f-0589-4760-8fae-c592a25784aa",
   "metadata": {},
   "source": [
    "#### C4\n",
    "\n",
    "主要是computational gene sets，包括cancer cell, cancer gene, cancer modules，\n",
    "\n",
    "意义不是很大"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d401ce3f-c05d-4016-bf86-51278ef9f521",
   "metadata": {},
   "outputs": [],
   "source": [
    "b = msigdbr(species = \"Homo sapiens\", category = \"C4\") %>%\n",
    "  dplyr::select(gs_name, entrez_gene)\n",
    "saveRDS(b,'gsea_c4.rds')\n",
    "\n",
    "gsea_results=lapply(\n",
    "  compounds,\n",
    "  FUN=function(x){\n",
    "    print(x)\n",
    "    gs=out %>% filter(compound==x) %>% filter(p_val_adj<0.05) %>% dplyr::select(c('GeneSymbol','avg_logFC'))\n",
    "    \n",
    "    if(nrow(gs)>1){ #有些组可能相比DMSO，没有差异基因\n",
    "      a1 = gs$avg_logFC\n",
    "      a2 = easyConvert(\n",
    "        species = \"HUMAN\",\n",
    "        queryList = gs$GeneSymbol,\n",
    "        queryType = \"SYMBOL\"\n",
    "      )\n",
    "      names(a1) = a2$ENTREZID\n",
    "      b1 = a1[order(a1, decreasing = T)]\n",
    "      \n",
    "      gsearesults = GSEA(b1, TERM2GENE = b,pvalueCutoff=1)\n",
    "      return(gsearesults)\n",
    "    }else{\n",
    "      gsearesults=list()\n",
    "      return(gsearesults)\n",
    "    }\n",
    "  }\n",
    ")\n",
    "names(gsea_results)=compounds\n",
    "#saveRDS(gsea_results,'gsea_results.rds')\n",
    "saveRDS(gsea_results,'gsea_c4_results.rds')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "458f1f00-eb0e-4292-8a64-24ebbd882075",
   "metadata": {},
   "source": [
    "+ 结果可视化"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5f9ec85-d233-42dd-8798-eb4615e81ac5",
   "metadata": {},
   "outputs": [],
   "source": [
    "ps_gseas=list()\n",
    "ps_gseas=lapply(\n",
    "  compounds,\n",
    "  FUN = function(x){\n",
    "    print(x)\n",
    "    tmp=gsea_results[[x]]\n",
    "    if(length(tmp)){\n",
    "      p=tmp@result %>% mutate(Description = stringr::str_to_title(gsub(\"_\", \" \", Description))) %>%\n",
    "        mutate(Description = reorder(Description, setSize)) %>% head(10) %>%\n",
    "        ggplot(aes(\n",
    "          x = enrichmentScore,\n",
    "          y = Description,\n",
    "          size = setSize,\n",
    "          color = pvalue\n",
    "        )) +\n",
    "        geom_point() +\n",
    "        scale_size(range = c(1, 5), name = \"Size\") +\n",
    "        scale_color_gradient(low = \"blue\", high = \"red\") +\n",
    "        xlab(\"EScore\") + ylab(\"\") + labs(title=x)+\n",
    "        theme(axis.text.y = element_text(size = 10, color = \"black\"))+\n",
    "        scale_y_discrete(labels = function(x) stringr::str_wrap(x, width = 25))+\n",
    "        theme_bw(base_size = 15)\n",
    "      \n",
    "    }else{\n",
    "      p=ggplot()\n",
    "    }\n",
    "    return(p)\n",
    "  }\n",
    ")\n",
    "\n",
    "#保存\n",
    "ggsave(\n",
    "  filename = \"GSEA_c4_plots.pdf\", \n",
    "  plot = gridExtra::marrangeGrob(ps_gseas, nrow=1, ncol=2), \n",
    "  width = 12, height = 6\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c47fd64b-7439-4c1f-bf59-e559318705c0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2b3d2f9a-44e4-49cb-84cc-addbb7c70e18",
   "metadata": {},
   "source": [
    "#### C5\n",
    "\n",
    "ontology gene set，主要是:\n",
    "\n",
    "+ BP\n",
    "+ CC\n",
    "+ MF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c785f25-676a-4960-bf60-7ca639508ec9",
   "metadata": {},
   "outputs": [],
   "source": [
    "C5=msigdbr(species = \"Homo sapiens\", category = \"C5\") %>%\n",
    "  dplyr::select(gs_name, entrez_gene)\n",
    "saveRDS(C5,'gsea_c5.rds')\n",
    "b=C5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f30e65a7-fedf-48dc-a7ec-a162cd90c2a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "C5=msigdbr(species = \"Homo sapiens\", category = \"C5\") %>%\n",
    "  dplyr::select(gs_name, entrez_gene)\n",
    "saveRDS(C5,'gsea_c5.rds')\n",
    "b=C5\n",
    "\n",
    "gsea_results=lapply(\n",
    "  compounds,\n",
    "  FUN=function(x){\n",
    "    print(x)\n",
    "    gs=out %>% filter(compound==x) %>% filter(p_val_adj<0.05) %>% dplyr::select(c('GeneSymbol','avg_logFC'))\n",
    "    \n",
    "    if(nrow(gs)>1){ #有些组可能相比DMSO，没有差异基因\n",
    "      a1 = gs$avg_logFC\n",
    "      a2 = easyConvert(\n",
    "        species = \"HUMAN\",\n",
    "        queryList = gs$GeneSymbol,\n",
    "        queryType = \"SYMBOL\"\n",
    "      )\n",
    "      names(a1) = a2$ENTREZID\n",
    "      b1 = a1[order(a1, decreasing = T)]\n",
    "      \n",
    "      gsearesults = GSEA(b1, TERM2GENE = b,pvalueCutoff=1)\n",
    "      return(gsearesults)\n",
    "    }else{\n",
    "      gsearesults=list()\n",
    "      return(gsearesults)\n",
    "    }\n",
    "  }\n",
    ")\n",
    "names(gsea_results)=compounds\n",
    "#saveRDS(gsea_results,'gsea_results.rds')\n",
    "saveRDS(gsea_results,'gsea_c5_results.rds')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be0bcac4-c30d-4300-8a67-14d60d9100d3",
   "metadata": {},
   "source": [
    "+ 结果可视化"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d7570b8-6604-41ca-94fb-1dc5da34ee67",
   "metadata": {},
   "outputs": [],
   "source": [
    "ps_gseas=list()\n",
    "ps_gseas=lapply(\n",
    "  compounds,\n",
    "  FUN = function(x){\n",
    "    print(x)\n",
    "    tmp=gsea_results[[x]]\n",
    "    if(length(tmp)){\n",
    "      p=tmp@result %>% mutate(Description = stringr::str_to_title(gsub(\"_\", \" \", Description))) %>%\n",
    "        mutate(Description = reorder(Description, setSize)) %>% head(10) %>%\n",
    "        ggplot(aes(\n",
    "          x = enrichmentScore,\n",
    "          y = Description,\n",
    "          size = setSize,\n",
    "          color = pvalue\n",
    "        )) +\n",
    "        geom_point() +\n",
    "        scale_size(range = c(1, 5), name = \"Size\") +\n",
    "        scale_color_gradient(low = \"blue\", high = \"red\") +\n",
    "        xlab(\"EScore\") + ylab(\"\") + labs(title=x)+\n",
    "        theme(axis.text.y = element_text(size = 10, color = \"black\"))+\n",
    "        scale_y_discrete(labels = function(x) stringr::str_wrap(x, width = 25))+\n",
    "        theme_bw(base_size = 15)\n",
    "      \n",
    "    }else{\n",
    "      p=ggplot()\n",
    "    }\n",
    "    return(p)\n",
    "  }\n",
    ")\n",
    "\n",
    "#保存\n",
    "ggsave(\n",
    "  filename = \"GSEA_c5_plots.pdf\", \n",
    "  plot = gridExtra::marrangeGrob(ps_gseas, nrow=1, ncol=2), \n",
    "  width = 12, height = 6\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "210f6677-073c-43f0-8e3d-96fbf798ffe0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "887103df-89b7-44d6-8b0f-44061a25dc01",
   "metadata": {},
   "source": [
    "#### C6\n",
    "\n",
    "oncogenic signature gene sets，一些研究相关的数据，比较有意义，可以找到targets，pathways perturbation等"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "034d34bf-5354-438f-9566-64db972c0505",
   "metadata": {},
   "outputs": [],
   "source": [
    "C6=msigdbr(species = \"Homo sapiens\", category = \"C6\") %>%\n",
    "  dplyr::select(gs_name, entrez_gene)\n",
    "saveRDS(C6,'gsea_c6.rds')\n",
    "b=C5\n",
    "\n",
    "\n",
    "gsea_results=lapply(\n",
    "  compounds,\n",
    "  FUN=function(x){\n",
    "    print(x)\n",
    "    gs=out %>% filter(compound==x) %>% filter(p_val_adj<0.05) %>% dplyr::select(c('GeneSymbol','avg_logFC'))\n",
    "    \n",
    "    if(nrow(gs)>1){ #有些组可能相比DMSO，没有差异基因\n",
    "      a1 = gs$avg_logFC\n",
    "      a2 = easyConvert(\n",
    "        species = \"HUMAN\",\n",
    "        queryList = gs$GeneSymbol,\n",
    "        queryType = \"SYMBOL\"\n",
    "      )\n",
    "      names(a1) = a2$ENTREZID\n",
    "      b1 = a1[order(a1, decreasing = T)]\n",
    "      \n",
    "      gsearesults = GSEA(b1, TERM2GENE = b,pvalueCutoff=1)\n",
    "      return(gsearesults)\n",
    "    }else{\n",
    "      gsearesults=list()\n",
    "      return(gsearesults)\n",
    "    }\n",
    "  }\n",
    ")\n",
    "names(gsea_results)=compounds\n",
    "#saveRDS(gsea_results,'gsea_results.rds')\n",
    "saveRDS(gsea_results,'gsea_c6_results.rds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80a40532-b063-426a-ac0a-5427a26c4013",
   "metadata": {},
   "outputs": [],
   "source": [
    "ps_gseas=list()\n",
    "ps_gseas=lapply(\n",
    "  compounds,\n",
    "  FUN = function(x){\n",
    "    print(x)\n",
    "    tmp=gsea_results[[x]]\n",
    "    if(length(tmp)){\n",
    "      p=tmp@result %>% mutate(Description = stringr::str_to_title(gsub(\"_\", \" \", Description))) %>%\n",
    "        mutate(Description = reorder(Description, setSize)) %>% head(10) %>%\n",
    "        ggplot(aes(\n",
    "          x = enrichmentScore,\n",
    "          y = Description,\n",
    "          size = setSize,\n",
    "          color = pvalue\n",
    "        )) +\n",
    "        geom_point() +\n",
    "        scale_size(range = c(1, 5), name = \"Size\") +\n",
    "        scale_color_gradient(low = \"blue\", high = \"red\") +\n",
    "        xlab(\"EScore\") + ylab(\"\") + labs(title=x)+\n",
    "        theme(axis.text.y = element_text(size = 10, color = \"black\"))+\n",
    "        scale_y_discrete(labels = function(x) stringr::str_wrap(x, width = 25))+\n",
    "        theme_bw(base_size = 15)\n",
    "      \n",
    "    }else{\n",
    "      p=ggplot()\n",
    "    }\n",
    "    return(p)\n",
    "  }\n",
    ")\n",
    "\n",
    "#保存\n",
    "ggsave(\n",
    "  filename = \"GSEA_c5_plots.pdf\", \n",
    "  plot = gridExtra::marrangeGrob(ps_gseas, nrow=1, ncol=2), \n",
    "  width = 12, height = 6\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d881c1b6-3633-4ccc-bf63-252a097d7b9d",
   "metadata": {},
   "source": [
    "#### GSEA plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b18c4e9-4d2c-42a8-9d7f-61825a3511d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_results=readRDS('gsea_c6_results.rds')\n",
    "\n",
    "#对个别GSEA结果进行可视化\n",
    "#https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html\n",
    "enrichplot::gseaplot2(gsea_results$`(S)-Crizotinib`, \n",
    "                      geneSetID = 1:3, pvalue_table = TRUE)\n",
    "\n",
    "enrichplot::gseaplot2(gsea_results$AZD8055, \n",
    "                      geneSetID = 1:3, pvalue_table = TRUE)\n",
    "enrichplot::gseaplot2(gsea_results$Cmp_334, \n",
    "                      geneSetID = 1:3, pvalue_table = TRUE)\n",
    "\n",
    "\n",
    "enrichplot::gseaplot2(gsea_results$Brusatol, \n",
    "                      geneSetID = 1:3, pvalue_table = TRUE)\n",
    "enrichplot::gseaplot2(gsea_results$Homoharringtonine, \n",
    "                      geneSetID = 1:3, pvalue_table = TRUE)\n",
    "enrichplot::gseaplot2(gsea_results$Triptolide, \n",
    "                      geneSetID = 1:3, pvalue_table = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5627455-6ba6-48b5-817d-7a56a9f7fd9e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04d0250c-fdfe-42da-8717-02c7a2ab45ea",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c83555d2-9c39-4a02-ae06-b28a2b32e6e3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44b13818-5625-4b10-bd75-4cfb06b51d81",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R361",
   "language": "R",
   "name": "ir361"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
